TELCODOCS#2488: Add GNR-D T-BC documentation (#111476)
openshift/openshift-docs: modules/nw-ptp-gnrd-t-bc-holdover.adoc

// Module included in the following assemblies:
//
// * networking/ptp/configuring-ptp.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-ptp-gnrd-t-bc-holdover_{context}"]
= Configuring GNR-D T-BC holdover on a GNR-D platform

[role="_abstract"]
You can configure a Intel® Granite Rapids-D (GNR-D) platform device as telecom boundary clock (T-BC) with holdover support by using the PTP Operator.

In this configuration, one time receiver (TR) port synchronizes to an upstream telecom grandmaster clock, while time transmitter (TT) ports distribute synchronized time to downstream devices. If the upstream timing source degrades, disconnects, or becomes unavailable, the system enters holdover mode and maintains timing by using configured digital phase-locked loop (DPLL) devices.

This configuration requires path-based network interface naming and uses the following custom resources:

> * `MachineConfig`

Reviewer: @sr1kar99 @vitus133 Machine Config introduced for the first time in the PTP Docs, what's the best practice here? I personally would like to know 'Why' I need this. and what it does. Can you ref the Machine Config docs for 'Further reading'?

Author: Updated as per the suggestion.

* `PtpConfig`

> * `HardwareConfig`

Reviewer: @sr1kar99 @vitus133  Ditto. New configuration, and a PTP only CR, no introduction, or refs for this CR API responsibility in the composition.

Author: Updated as per the suggestion.

    ts2phcOpts: -s generic -a --ts2phc.rh_external_pps 1
  recommend:
  - match:
    - nodeLabel: node-role.kubernetes.io/master
    priority: 4
    profile: 00-tbc-tt
  - match:
    - nodeLabel: node-role.kubernetes.io/master
    priority: 4
    profile: 01-tbc-tr
----
+
* The TT profile configures downstream-facing ports.
* The TR profile configures the upstream-facing port, `phc2sys`, and `ts2phc`.
* `controllingProfile` links the TT profile to the TR profile.
* `clockType: T-BC` enables telecom boundary clock behavior.

. Apply the `PtpConfig` CR by running the following command:
+
[source,terminal]
----
$ oc apply -f t-bc-config.yaml
----

. Create a `HardwareConfig` CR to configure holdover thresholds:
+
[source,yaml]
----
apiVersion: ptp.openshift.io/v2alpha1
kind: HardwareConfig
metadata:
  name: t-bc
spec:
  profile:
    name: GNR-D-T-BC
    clockType: T-BC
    clockChain:
      structure:
      - name: leader
        hardwareSpecificDefinitions: <gnrd_platform>
        dpll:
          holdoverParameters:
            maxInSpecOffset: 40
            localMaxHoldoverOffset: 1500
            localHoldoverTimeout: 14400
  relatedPtpProfileName: 01-tbc-tr
----
+

> * `hardwareSpecificDefinitions: <gnrd_platform>` applies hardware-specific behavior of the GNR-D platform. Replace `<gnrd_platform>` with the GNR-D platform on which you want to configure T-BC holdover.

Reviewer 2: Can we add that there are two possible options here at the moment: ` dell/XR8720t` or ` hpe/EL140-Gen12`?

Author: Updated it as follows:
"* `hardwareSpecificDefinitions: <gnrd_platform>` applies hardware-specific behavior of the GNR-D platform. Replace `<gnrd_platform>` with the GNR-D platform on which you want to configure T-BC holdover. The supported GNR-D platforms are `dell/XR8720t` and  `hpe/EL140-Gen12`."

// Module included in the following assemblies:
//
// * networking/ptp/configuring-ptp.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-ptp-gnrd-t-bc-holdover_{context}"]
= Configuring GNR-D T-BC holdover on Dell XR8720t

[role="_abstract"]

> You can configure a Dell XR8720t platform as a Global Navigation Receiver Device (GNR-D) telecom boundary clock (T-BC) with holdover support by using the PTP Operator.

Reviewer 2: ```suggestion
You can configure a "Intel® Granite Rapids-D" (GNR-D) platform device as telecom boundary clock (T-BC) with holdover support by using the PTP Operator.
```

Author: Updated as per the suggestion.

> In this configuration, one time receiver (TR) port synchronizes to an upstream telecom grandmaster clock, while time transmitter (TT) ports distribute synchronized time to downstream devices. If the upstream timing source degrades, disconnects, or becomes unavailable, the system enters holdover mode and maintains timing by using configured digital phase-locked loop (DPLL) thresholds.

Reviewer 2: ```suggestion
In this configuration, one time receiver (TR) port synchronizes to an upstream telecom grandmaster clock, while time transmitter (TT) ports distribute synchronized time to downstream devices. If the upstream timing source degrades, disconnects, or becomes unavailable, the system enters holdover mode and maintains timing by using configured digital phase-locked loop (DPLL) devices.
```

Author: Updated as per the suggestion.

This configuration requires path-based network interface naming and uses the following custom resources:

* `MachineConfig`
* `PtpConfig`
* `HardwareConfig`

.Prerequisites

* Install the OpenShift CLI (`oc`).
* Log in as a user with `cluster-admin` privileges.
* Install the PTP Operator.

> * Deploy a Dell XR8720t platform with supported NIC hardware.

Reviewer 2: I'm not sure we should explicitly target this document to the Dell XR8720t platform. Other platforms can also be supported with the only difference in the `HardwareConfig` CR applied:
`hardwareSpecificDefinitions: dell/XR8720t`
will be something else on other platforms. 
Do you think we can (should?) be more generic here?

Author: Agreed! 
Updated it as follows:

* `hardwareSpecificDefinitions: <gnrd_platform>` applies hardware-specific behavior of the GNR-D platform. Replace `<gnrd_platform>` with the GNR-D platform on which you want to configure T-BC holdover.

WDYT?

* Identify the TR and TT interfaces for your hardware layout.

.Procedure

> . Configure path-based network interface naming for the Dell XR8720t platform:

Reviewer 2: Dell XR8720t - same question as above

Author: Updated it to be generic as per the suggestion.

      #
      # Port Data Set
      #
      logAnnounceInterval -3
      logSyncInterval -4
      logMinDelayReqInterval -4
      logMinPdelayReqInterval -4
      announceReceiptTimeout 3
      syncReceiptTimeout 0
      delayAsymmetry 0
      fault_reset_interval -4
      neighborPropDelayThresh 20000000
      masterOnly 0
      G.8275.portDS.localPriority 128
      #
      # Run time options
      #
      assume_two_step 0
      logging_level 6
      path_trace_enabled 0
      follow_up_info 0
      hybrid_e2e 0
      inhibit_multicast_service 0
      net_sync_monitor 0
      tc_spanning_tree 0
      tx_timestamp_timeout 50
      unicast_listen 0
      unicast_master_table 0
      unicast_req_duration 3600
      use_syslog 1
      verbose 0
      summary_interval 0
      kernel_leap 1
      check_fup_sync 0
      clock_class_threshold 135
      #
      # Servo Options
      #
      pi_proportional_const 0.60
      pi_integral_const 0.001
      pi_proportional_scale 0.0
      pi_proportional_exponent -0.3
      pi_proportional_norm_max 0.7
      pi_integral_scale 0.0
      pi_integral_exponent 0.4
      pi_integral_norm_max 0.3
      step_threshold 2.0

>       # end T-BC experiment

Reviewer 2: ```suggestion
```

      #
      # Port Data Set
      #
      logAnnounceInterval -3
      logSyncInterval -4
      logMinDelayReqInterval -4
      logMinPdelayReqInterval -4
      announceReceiptTimeout 3
      syncReceiptTimeout 0
      delayAsymmetry 0
      fault_reset_interval -4
      neighborPropDelayThresh 20000000
      masterOnly 0
      G.8275.portDS.localPriority 128
      #
      # Run time options
      #
      assume_two_step 0
      logging_level 6
      path_trace_enabled 0
      follow_up_info 0
      hybrid_e2e 0
      inhibit_multicast_service 0
      net_sync_monitor 0
      tc_spanning_tree 0
      tx_timestamp_timeout 50
      unicast_listen 0
      unicast_master_table 0
      unicast_req_duration 3600
      use_syslog 1
      verbose 0
      summary_interval 0
      kernel_leap 1
      check_fup_sync 0
      clock_class_threshold 135
      #
      # Servo Options
      #
      pi_proportional_const 0.60
      pi_integral_const 0.001
      pi_proportional_scale 0.0
      pi_proportional_exponent -0.3
      pi_proportional_norm_max 0.7
      pi_integral_scale 0.0
      pi_integral_exponent 0.4
      pi_integral_norm_max 0.3
      step_threshold 2.0

>       # end T-BC experiment

Reviewer 2: ```suggestion
```

      ts2phc.channel 0
      ts2phc.pin_index 1
    ts2phcOpts: -s generic -a --ts2phc.rh_external_pps 1
  recommend:
  - match:
    - nodeLabel: node-role.kubernetes.io/master
    priority: 4
    profile: 00-tbc-tt
  - match:
    - nodeLabel: node-role.kubernetes.io/master
    priority: 4
    profile: 01-tbc-tr
----
+
* The TT profile configures downstream-facing ports.
* The TR profile configures the upstream-facing port, `phc2sys`, and `ts2phc`.
* `controllingProfile` links the TT profile to the TR profile.
* `clockType: T-BC` enables telecom boundary clock behavior.

. Apply the `PtpConfig` CR by running the following command:
+
[source,terminal]
----
$ oc apply -f t-bc-config.yaml
----

. Create a `HardwareConfig` CR to configure holdover thresholds:
+
[source,yaml]
----
apiVersion: ptp.openshift.io/v2alpha1
kind: HardwareConfig
metadata:
  name: t-bc
spec:
  profile:
    name: GNR-D-T-BC
    clockType: T-BC
    clockChain:
      structure:
      - name: leader
        hardwareSpecificDefinitions: dell/XR8720t
        dpll:
          holdoverParameters:
            maxInSpecOffset: 40
            localMaxHoldoverOffset: 1500
            localHoldoverTimeout: 14400

>   relatedPtpProfileName: 01-tbc-tr

Reviewer 2: We are still discussing this line, there is a possibility we'll need to make a change or add a clarification

----
+
* `hardwareSpecificDefinitions: <gnrd_platform>` applies hardware-specific behavior of the GNR-D platform. Replace `<gnrd_platform>` with the GNR-D platform on which you want to configure T-BC holdover. The supported GNR-D platforms are `dell/XR8720t` and  `hpe/EL140-Gen12`.

> * `holdoverParameters` configure DPLL holdover thresholds.

Reviewer 3: Wondering if it shouldn't be `holdoverParameters` configure**s** as it's the name of a field. Either way is probably fine and yours flows better!

* `relatedPtpProfileName` links the hardware profile to the TR profile.

. Apply the `HardwareConfig` CR by running the following command:
+
[source,terminal]
----
$ oc apply -f hardwareconfig.yaml
----

.Verification

* Verify that the `PtpConfig` and `HardwareConfig` resources are deployed:
+
[source,terminal]
----
$ oc get ptpconfig -n openshift-ptp
$ oc get hardwareconfig
----

* Check the `linuxptp-daemon` logs:
+
[source,terminal]
----
$ oc logs ds/linuxptp-daemon -n openshift-ptp -c linuxptp-daemon-container --since=1s -f | grep T-BC
----
+
.Example output
[source,terminal]
----
T-BC[1760525446]:[ts2phc.1.config] eno8703 offset 1 T-BC-STATUS s2
T-BC[1760525447]:[ts2phc.1.config] eno8703 offset 1 T-BC-STATUS s1
----
+
The status is reported every second:
+

> * `s2`: Locked - synchronized with the upstream clock

Reviewer 3: You need ** rather than * for these three, to push the bullets in a level.

Author: great catch! Updated as per the suggestion.
